### PR TITLE
[feat] add CosFace paper's LMCL to MEVO

### LINKED ISSUE #864: [feature] support normalization, scale and margin in MEVO
## 🚀 Feature
currently, vanilla FC + softmax + nllloss is supported in MEVO. In face recognition, cosface loss is commonly used which additionally normalize both the input feature and fc weights and add scale and margin to the logits.

## Motivation
face recognition models are typically trained with millions of IDs. MEVO will have broader impact in face recognition field if CosFace is supported.

## Additional context
CosFace paper: https://arxiv.org/abs/1801.09414

### PR DESCRIPTION
**Pull request by min-xu-ai**
- added baseline algorithm to the reference kernel
- added MEVO version of LMCL
- added unit test to verify it is correct with respect to the reference as well as its memory usage

## What does this PR do?
Fixes #864 .

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Review by tmarkstrum — APPROVED** [review]
Please resolve the two comments. Other than that, LGTM.

File: fairscale/experimental/nn/mevo.py [hunk 1]
```diff
@@ -77,10 +81,23 @@ class BaselineSoftmax(nn.Module):
...
+            Used in LMCL. If scale is None, LMCL is turned off. See
+            MEVO comments for more details.
 
     """
 
-    def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log_softmax: bool = True):
+    def __init__(
+        self,
+        proj_weight: nn.Parameter,
+        tile_factor: int = 0,
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:96]: right, this part is to make the API uniform for better testing. :-)

File: fairscale/experimental/nn/mevo.py [hunk 2]
```diff
@@ -92,6 +109,27 @@ def __init__(self, proj_weight: nn.Parameter, tile_factor: int = 0, log_softmax:
...
+        #   k: number of classes (vocab size)
+        #   c: hidden dimension (d_model)
+        x = F.normalize(input, dim=1)
+        w = F.normalize(self.fc.weight, dim=1)
+        logits = torch.einsum("nc,kc->nk", x, w)
+
+        # add margin
+        row_ind = torch.arange(x.shape[0], dtype=torch.long).to(x.device)
+        col_ind = target
+        logits[row_ind, col_ind] -= self.margin
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply fairscale/experimental/nn/mevo.py:127]: Yeah, that's not idea. But I took it from the original sample code in the issue provided by the reporter. I think by "adding", we just mean adding this operation.
